File: leetcode/problems/c/relative-ranks.cpp
// by @codeAbinash
// Time : O(n * log(n))
// Space : O(n)

#include <queue>
#include <string>
#include <vector>

using namespace std;

// Using priority_queue
class Solution {
 public:
   vector<string> findRelativeRanks(vector<int>& score) {
      priority_queue<pair<int, int>> pq;
      for (int i = 0; i < score.size(); ++i) pq.push({score[i], i});
      vector<string> res(score.size());
      for (int i = 0; i < score.size(); ++i) {
         if (i == 0)
            res[pq.top().second] = "Gold Medal";
         else if (i == 1)
            res[pq.top().second] = "Silver Medal";
         else if (i == 2)
            res[pq.top().second] = "Bronze Medal";
         else
            res[pq.top().second] = to_string(i + 1);
         pq.pop();
      }
      return res;
   }
};

// Using sort
class Solution2 {
 public:
   vector<string> findRelativeRanks(vector<int>& score) {
      vector<pair<int, int>> v;
      for (int i = 0; i < score.size(); ++i) v.push_back({score[i], i});
      sort(v.rbegin(), v.rend());
      vector<string> res(score.size());
      for (int i = 0; i < score.size(); ++i) {
         if (i == 0)
            res[v[i].second] = "Gold Medal";
         else if (i == 1)
            res[v[i].second] = "Silver Medal";
         else if (i == 2)
            res[v[i].second] = "Bronze Medal";
         else
            res[v[i].second] = to_string(i + 1);
      }
      return res;
   }
};

File: leetcode/problems/java/relative-ranks.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {
  public String[] findRelativeRanks(int[] score) {
    Integer[] sort = new Integer[score.length];
    for (int i = 0; i < score.length; i++) {
      sort[i] = i;
    }

    Arrays.sort(sort, (a, b) -> (score[b] - score[a]));

    String[] result = new String[score.length];

    for (int i = 0; i < score.length; i++) {
      if (i == 0) {
        result[sort[i]] = "Gold Medal";
      } else if (i == 1) {
        result[sort[i]] = "Silver Medal";
      } else if (i == 2) {
        result[sort[i]] = "Bronze Medal";
      } else {
        result[sort[i]] = String.valueOf(i + 1);
      }
    }

    return result;
  }
}

class Solution2 {
  // Using priority queue
  public String[] findRelativeRanks(int[] score) {
    int n = score.length;
    String[] result = new String[n];
    PriorityQueue<Integer> pq = new PriorityQueue<>(n, (a, b) -> score[b] - score[a]);

    for (int i = 0; i < n; i++) {
      pq.add(i);
    }

    for (int i = 0; i < n; i++) {
      int index = pq.poll();
      if (i == 0) {
        result[index] = "Gold Medal";
      } else if (i == 1) {
        result[index] = "Silver Medal";
      } else if (i == 2) {
        result[index] = "Bronze Medal";
      } else {
        result[index] = String.valueOf(i + 1);
      }
    }

    return result;
  }
}